### DISCUSSION TITLE
[help] How to handle mixed local and cloud storage

### DISCUSSION THREAD

**Discussion by tarensanders** [ropensci/targets#1381] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I've been using cloud storage with great success, but ran into an issue when using a mix of cloud storage and local storage when trying to move between machines.

I'm using mixed storage methods to try to reduce the cost of cloud storage. I have a few targets that take a very long time to generate which get sent to the cloud, and then some much faster targets (some of which generate large objects) which make sense to keep locally rather than have to pay for them to be checked on each pipeline run. I'm using GCS, so as a sketch the pipeline looks something like:

```R
library(targets)
library(tarchetypes)

tar_option_set(
  resources = tar_resources(
    gcp = tar_resources_gcp(bucket = "mybucket", prefix = "mybucket")
  )
)

list(
  tar_target(slow_target, foo(), repository = "gcp"), 
  tar_target(fast_target, bar(slow_target))
)
```

I then keep `_targets/meta/meta` under version control. This works great, but the issue is when I then go to setup another computer. I needed to use a HPC to rerun the slow targets so I cloned the repo and tried to rerun the pipeline, but I ran into an error that targets was trying to access a file that doesn't exist. In this example, targets would be looking for `_targets/objects/fast_target`, but because `/objects` doesn't exist this fails. In the end I just ended up copying in the local objects to resolve the issue, but that seems hacky.

In retrospect I probably should have used `tar_invalidate` on the local targets. But is there any other way to deal with this?

**Comment by Aariq** [comment]
What about instead of syncing _targets/meta/meta with version control, use `tar_meta_sync()`?  I'm not sure how that handles non-cloud targets, but maybe it already handles this issue?

**Reply by tarensanders** [reply]
That's a good suggestion. The [function description](https://docs.ropensci.org/targets/reference/tar_meta_sync.html) doesn't address this case:

> For each metadata file, if the local file does not exist or is older than the cloud file, then the cloud file is downloaded to the local file path. Conversely, if the cloud file is older or does not exist, then the local file is uploaded to the cloud.

Looking at the [source code](https://github.com/ropensci/targets/blob/main/R/class_database.R) for the sync function, it doesn't look like this would be handled.

**Comment by wlandau** [comment]
If you're trying to copy over the pipeline to another machine and that pipeline depends on local files in `_targets/objects`, then those local files would need to be copied over too. I understand your reluctance to upload them to the cloud if they are large, but I don't see a way around this tradeoff.

**Reply by tarensanders** [reply]
Thanks, @wlandau. I'm totally fine with that as a workaround, or using `tar_invalidate()` to force the rebuild. I think the thing that trips me up is that it was not immediately obvious (to me) why this was happening. Would you be interested in a PR that checks that local files in `_targets/objects` exists and gives a meaningful error message?

**Reply by wlandau** [reply]
> I ran into an error that targets was trying to access a file that doesn't exist.

During `tar_read()`? Because `tar_make()` will just rerun the target if the file does not exist.

```r
# _targets.R file
library(targets)
list(
  tar_target(a, 1),
  tar_target(b, 2)
)
```

```r
tar_make(reporter = "silent")
unlink("_targets/objects", recursive = TRUE)
```

```r
tar_read(a)
#> Error in gzfile(file, "rb") : cannot open the connection
#> In addition: Warning message:
#> In gzfile(file, "rb") :
#>   cannot open compressed file '_targets/objects/a', probable reason 'No such file or directory'
```

```r
tar_make()
#> ▶ dispatched target a
#> ● completed target a [0.001 seconds, 50 bytes]
#> ▶ dispatched target b
#> ● completed target b [0 seconds, 49 bytes]
#> ▶ ended pipeline [0.1 seconds]
```

Your case seems similar because `fast_target` is not a dependency of anything else, and presumably `slow_target` on GCP is still accessible on the new machine.

> Would you be interested in a PR that checks that local files in _targets/objects exists and gives a meaningful error message?

I think the `tar_read()` error message is clear enough. If your scenario is different, it would help to learn more.

**Reply by tarensanders** [reply]
Thanks, @wlandau. I was sure that I was getting the read error when running `tar_make()` but I can't make a reprex that does that.
